diff -r 19fc7167c67f src/pkg/flag/flag.go
--- a/src/pkg/flag/flag.go      Sun Apr 20 12:26:05 2014 +0200
+++ b/src/pkg/flag/flag.go      Mon May 05 18:27:37 2014 +0200
@@ -828,7 +828,7 @@
 // CommandLine is the default set of command-line flags, parsed from os.Args.
 // The top-level functions such as BoolVar, Arg, and on are wrappers for the
 // methods of CommandLine.
-var CommandLine = NewFlagSet(os.Args[0], ExitOnError)
+var CommandLine = NewFlagSet("", ExitOnError)

 // NewFlagSet returns a new, empty flag set with the specified name and
 // error handling property.
